Feature/imanol/iss 32 api libros crud

Imamtz0104 opened the pull request:
Resuelve #<ID>
 Resuelve #32

**Resumen**
- Implementa CRUD completo para la API de libros
- Corrige problemas de duplicación de código del PR #40
- Usa bookDb centralizado de db.ts (no duplicación)
- Formato de respuesta consistente con otros endpoints

**Cómo probar**

```bash
# 1. Listar libros existentes
curl -X GET http://localhost:3000/api/books

# 2. Crear nuevo libro
curl -X POST http://localhost:3000/api/books \
  -H "Content-Type: application/json" \
  -d '{"titulo":"1984","autor":"George Orwell","isbn":"978-84-376-0496-1"}'

# 3. Obtener libro específico (usar ID del paso anterior)
curl -X GET http://localhost:3000/api/books/51c2bcd7-f586-4a6a-a65d-20a34c972111

# 4. Actualizar libro
curl -X PUT http://localhost:3000/api/books/51c2bcd7-f586-4a6a-a65d-20a34c972111 \
  -H "Content-Type: application/json" \
  -d '{"titulo":"1984 - Edición Especial","disponible":false}'

# 5. Eliminar libro
curl -X DELETE http://localhost:3000/api/books/51c2bcd7-f586-4a6a-a65d-20a34c972111

JahirJmnz committed:
feat(users): implementar API CRUD completa de usuarios #01

- GET /api/users - Listar todos los usuarios
- POST /api/users - Crear nuevo usuario con validaciones
- GET /api/users/[id] - Obtener usuario por ID
- PUT /api/users/[id] - Actualizar usuario
- DELETE /api/users/[id] - Eliminar usuario

Validaciones implementadas:
- Email único y formato válido
- Nombre mínimo 2 caracteres
- UUID válido para IDs
- Códigos de error apropiados (400, 404, 409, 500)

Tests: 7/7 tests de usuarios pasando ✅

JahirJmnz committed:
fix(users): corregir params await para Next.js 15 #01

- Actualizar tipos de params a Promise<{ id: string }>
- Añadir await antes de usar params.id
- Compatible con Next.js 15+

JahirJmnz committed:
docs: actualizar READMEs con progreso de API usuarios #01

- Actualizar READMEs de compañeros con estado del proyecto
- Marcar ISS-01 como completado en documentación
- Añadir contexto sobre PR #35 y próximos pasos
- Actualizar README principal con progreso del proyecto

JahirJmnz committed:
Merge pull request #35 from JahirJmnz/feature/jahir/ISS-01-api-usuarios

feat(users): implementar API CRUD completa de usuarios #01

Imamtz0104 committed:
feat(books): implementar CRUD completo de libros #32

- GET /api/books - Listar todos los libros
- POST /api/books - Crear nuevo libro con validaciones
- GET /api/books/[id] - Obtener libro por ID
- PUT /api/books/[id] - Actualizar libro
- DELETE /api/books/[id] - Eliminar libro

Usa bookDb de db.ts centralizado (no duplicaciÃ³n)
Formato de respuesta consistente con otros endpoints
Validaciones completas: tÃ­tulo, autor, ISBN Ãºnico
Manejo de errores 400, 404, 409, 500

Imamtz0104 committed:
docs(books): agregar documentaciÃ³n y pruebas para API de libros #32

- DocumentaciÃ³n completa con comandos curl
- Archivo de pruebas automatizadas test-books-api.js
- Ejemplos de validaciones y cÃ³digos de error
- GuÃ­a de uso para todos los endpoints

Imamtz0104 committed:
docs(pr): agregar plantilla completa del PR #32

- Referencia al issue #32
- Comandos curl de prueba con evidencia
- Checklist completo
- DocumentaciÃ³n de archivos modificados

JahirJmnz commented:
### ⚠️ PROBLEMA CRÍTICO: Modificación de bitácoras de otros integrantes

Hola @Imamtz0104,

He revisado el PR #42 y veo que **respondiste muy rápido** a mi review del PR #40 (solo 9 minutos después). ¡Eso es excelente! 👍

Sin embargo, hay un **problema crítico** que debe corregirse:

---

## 🚨 Problema: Modificaste bitácoras de TODOS los integrantes

Tu PR modifica estos archivos que NO deberías tocar:
- ❌ `BICATORA-INTEGRANTE/Antonio/README.md`
- ❌ `BICATORA-INTEGRANTE/Jahir/README.md`
- ❌ `BICATORA-INTEGRANTE/Jair/README.md`
- ✅ `BICATORA-INTEGRANTE/Imanol/README.md` (este SÍ puedes modificarlo)

### ¿Por qué es un problema?
1. Cada integrante debe mantener su propia bitácora
2. Si todos modificamos las bitácoras de otros, habrá conflictos constantes
3. Pierdes el crédito de tu trabajo individual

### ¿Cómo corregirlo?
```bash
# 1. Deshacer los cambios a las bitácoras de otros
git checkout main -- BICATORA-INTEGRANTE/Antonio/README.md
git checkout main -- BICATORA-INTEGRANTE/Jahir/README.md
git checkout main -- BICATORA-INTEGRANTE/Jair/README.md

# 2. Mantener solo tus cambios
git add BICATORA-INTEGRANTE/Imanol/README.md

# 3. Hacer commit
git commit -m "fix: revert changes to other members' logs"

# 4. Force push (en este caso está justificado)
git push origin feature/imanol/ISS-32-api-libros-crud --force
```

---

## 🚨 SEGUNDO PROBLEMA: Modificaste README.md principal

También modificaste `README.md` del proyecto. A menos que sea necesario para documentar la API de libros específicamente, este cambio debería ser un PR separado.

---

## 🚨 TERCER PROBLEMA: Añadiste PR-TEMPLATE.md

Veo que añadiste `PR-TEMPLATE.md`. Esto debería:
1. Ser un PR separado (no mezclado con la funcionalidad de libros)
2. Estar en `.github/pull_request_template.md` (ubicación estándar de GitHub)
3. Ser discutido con el equipo primero

---

## ✅ Lo que SÍ está bien en este PR:

1. ✅ **Corregiste el problema de duplicación** - Ya no usas `booksStore.ts`
2. ✅ **Usas `bookDb` correctamente** - Importas de `@/app/lib/db`
3. ✅ **Añadiste tests** - `test-books-api.js`
4. ✅ **Documentación de API** - `BOOKS-API-DOCUMENTATION.md` (aunque podría ser separado)
5. ✅ **Comandos curl completos** en el PR
6. ✅ **Referencia al issue #32**

---

## 🎯 Acción requerida:

### Opción A: Limpiar este PR (Recomendado)
1. Revertir cambios a bitácoras de otros (`Antonio`, `Jahir`, `Jair`)
2. Revertir cambios a `README.md` y `PR-TEMPLATE.md`
3. Dejar solo:
   - `app/api/books/route.ts`
   - `app/api/books/[id]/route.ts`
   - `BICATORA-INTEGRANTE/Imanol/README.md` (solo tu bitácora)
   - Opcionalmente: `test-books-api.js` y `BOOKS-API-DOCUMENTATION.md`

### Opción B: Separar en múltiples PRs
1. Este PR: Solo código de la API de libros
2. PR separado: Documentación (`BOOKS-API-DOCUMENTATION.md`, `README.md`)
3. PR separado: Template de PRs (`PR-TEMPLATE.md`)

---

## 📚 Workflow correcto:

**Regla de oro**: Un PR = Una funcionalidad

- ✅ PR para API de libros
- ✅ PR separado para documentación general
- ✅ PR separado para templates del proyecto

---

Por favor, corrige estos puntos y avísame cuando esté listo para re-revisar. El código funcional está muy bien, solo necesitamos limpiar los cambios no relacionados. 👍

Imamtz0104 committed:
fix(books): corregir validaciones y formato de respuestas #32

- Actualizar validaciÃ³n UUID_REGEX siguiendo patrÃ³n del proyecto
- Cambiar formato de errores de 'error' a 'message' para consistencia
- Mejorar mensajes de error siguiendo estÃ¡ndares del equipo
- Unificar formato de respuestas con otros endpoints

Imamtz0104 committed:
test(books): aÃ±adir tests completos y mejorar documentaciÃ³n #32

- Tests CRUD completo para libros en test-api.js
- Tests de validaciones especÃ­ficas (UUID, ISBN duplicado, 404)
- Mejorar descripciÃ³n del PR con checklist completo
- DocumentaciÃ³n actualizada con todos los endpoints

Imamtz0104 committed:
fix: revert changes to other members' logs and clean PR

- Revertir cambios a bitÃ¡coras de otros integrantes (Antonio, Jahir, Jair)
- Eliminar README.md principal (no relacionado con API de libros)
- Eliminar PR-TEMPLATE.md (deberÃ­a ser PR separado)
- Eliminar documentaciÃ³n y tests extras (mantener solo funcionalidad)
- Mantener solo: app/api/books/ y BICATORA-INTEGRANTE/Imanol/README.md

JahirJmnz commented:
### ⚠️ NUEVO PROBLEMA: Archivos de usuarios incluidos por error

Hola de nuevo @Imamtz0104,

Veo que limpiaste bien las bitácoras de otros integrantes. ¡Excelente! 👍

Sin embargo, hay un nuevo problema:

---

## 🚨 Problema: Estás añadiendo archivos de la API de usuarios

Tu PR #42 incluye estos archivos:
- ❌ `app/api/users/route.ts`
- ❌ `app/api/users/[id]/route.ts`

### ¿Por qué es un problema?

1. **No es tu tarea**: Tu issue #32 es para la API de **LIBROS**, no usuarios
2. **Los usuarios ya están en QA**: El PR #35 de Jahir mergeó la API de usuarios a la rama `qa`
3. **Tu PR va a main**: Estás intentando mergear a `main`, no a `qa`
4. **Duplicación de trabajo**: Jahir ya implementó la API de usuarios

### ¿Cómo pasó esto?

Probablemente:
1. Creaste tu rama desde `qa` en lugar de `main`
2. O hiciste merge de `qa` en tu rama
3. Por eso tienes los archivos de usuarios que ya están en `qa`

---

## 🔧 Cómo corregirlo:

### Opción 1: Revertir solo los archivos de users (Más rápido)

```bash
# Eliminar los archivos de users de tu rama
git rm app/api/users/route.ts
git rm app/api/users/[id]/route.ts

# Commit
git commit -m "fix: remove users API files (not part of this PR)"

# Push
git push origin feature/imanol/ISS-32-api-libros-crud
```

### Opción 2: Recrear la rama desde main limpio (Más limpio)

```bash
# Guardar tus cambios de books
git diff main -- app/api/books/ BICATORA-INTEGRANTE/Imanol/README.md test-api.js > books-changes.patch

# Crear nueva rama desde main limpio
git checkout main
git pull origin main
git checkout -b feature/imanol/ISS-32-api-libros-crud-v2

# Aplicar solo tus cambios de books
git apply books-changes.patch

# Push
git push origin feature/imanol/ISS-32-api-libros-crud-v2 --force

# Luego cerrar este PR y crear uno nuevo
```

---

## ✅ Archivos que DEBERÍAN estar en tu PR:

- ✅ `app/api/books/route.ts` (tu tarea)
- ✅ `app/api/books/[id]/route.ts` (tu tarea)
- ✅ `BICATORA-INTEGRANTE/Imanol/README.md` (tu bitácora)
- ✅ `test-api.js` (tests de libros)

## ❌ Archivos que NO deberían estar:

- ❌ `app/api/users/route.ts` (tarea de Jahir, ya en QA)
- ❌ `app/api/users/[id]/route.ts` (tarea de Jahir, ya en QA)

---

**Recomiendo Opción 1** (más rápido). Solo elimina los archivos de users y haz push.

Una vez corregido, estaré listo para aprobar el PR. El código de libros se ve bien. 👍

Imamtz0104 committed:
fix: remove users API files (not part of this PR)

- Eliminar app/api/users/route.ts (tarea de Jahir, ya en QA)
- Eliminar app/api/users/[id]/route.ts (tarea de Jahir, ya en QA)
- Mantener solo archivos relacionados con API de libros
- PR ahora contiene Ãºnicamente funcionalidad de libros

Imamtz0104 committed:
fix: remove all non-books API files from PR

- Revertir app/api/loans/ (no es mi tarea)
- Revertir test-api.js (no es mi tarea)
- Revertir .cursor/rules/rules.mdc (no es mi tarea)
- Mantener solo: app/api/books/ y mi bitÃ¡cora
- PR ahora contiene Ãºnicamente API de libros
